### DISCUSSION TITLE
Questions on Custom Completers

### DISCUSSION THREAD

**Discussion by bkilrain** [ajaxorg/ace#4529] [Q&A] [answered]
1. What is the difference between the concepts of `snippets` and `completers`? 
2. Can someone share a link to the source code which defines the interface for a completer object? Are all valid completer properties referenced in [this file](https://github.com/ajaxorg/ace/blob/68ad610814ce662363082f4b17b42feca05898a5/lib/ace/autocomplete.js)?
3. How can I adjust the styles for the [popup](https://github.com/ajaxorg/ace/blob/master/lib/ace/autocomplete/popup.js#L355)? Just a CSS override or is there a knob for that somewhere? Specifically, I want to have it mount wider the 300px to accommodate longer name/meta values.

Loving the library so far... it's clear, from other implementations in the wild, ACE is super powerful - I just need a few hints to know where to find my answers in src.

**Comment by bkilrain** [comment]
Answering my own question for #1 -

Snippets API is detailed here: https://cloud9-sdk.readme.io/docs/snippets though I'm still fuzzy on how to apply a custom snippets file.

**Comment by nightwing** [comment, marked as answer]
To add custom snippets you can use something similar to what  [language_tools does](https://github.com/ajaxorg/ace/blob/68ad610814ce662363082f4b17b42feca05898a5/lib/ace/ext/language_tools.js#L140-L141)

```js
        var {snippetManager} = ace.require("ace/snippets")
        var snippets = snippetManager.parseSnippetFile(snippetText);

        snippetManager.register(snippets, "defaultLanguageName");
```

if you store snippets as an array of objects you can use JSON.parse instead of using `parseSnippetFile`.
